lib/ts/recipe/thirdpartyemailpassword/utils.ts

```diff
@@ -0,0 +1,411 @@
+import { defaultEmailValidator, defaultPasswordValidator } from "../emailpassword/utils";
+
+export function validateAndNormaliseUserInput(
+    recipeInstance: Recipe,
+    appInfo: NormalisedAppinfo,
+    config: TypeInput
+): TypeNormalisedInput {
+    validateTheStructureOfUserInput(
+        config,
+        InputSchema,
+        "thirdpartyemailpassword recipe",
+        recipeInstance.getRecipeId()
+    );
+
+    let sessionFeature = validateAndNormaliseSessionFeatureConfig(
+        recipeInstance,
+        appInfo,
+        config === undefined ? undefined : config.sessionFeature
+    );
+
+    let signUpFeature = validateAndNormaliseSignUpConfig(
+        recipeInstance,
+        appInfo,
+        config === undefined ? undefined : config.signUpFeature
+    );
+
+    let signInFeature = validateAndNormaliseSignInConfig(
+        recipeInstance,
+        appInfo,
+        config === undefined ? undefined : config.signInFeature
+    );
+
+    let resetPasswordUsingTokenFeature = config.resetPasswordUsingTokenFeature;
+
+    let providers = config.providers === undefined ? [] : config.providers;
+
+    let signOutFeature = validateAndNormaliseSignOutConfig(
+        recipeInstance,
+        appInfo,
+        config === undefined ? undefined : config.signOutFeature
+    );
+
+    let emailVerificationFeature = validateAndNormaliseEmailVerificationConfig(
+        recipeInstance,
+        appInfo,
+        config === undefined ? undefined : config.emailVerificationFeature
+    );
+
+    return {
+        sessionFeature,
+        signUpFeature,
+        signInFeature,
+        providers,
+        signOutFeature,
+        resetPasswordUsingTokenFeature,
+        emailVerificationFeature,
+    };
+}
+
+async function defaultValidator(value: any): Promise<string | undefined> {
+    return undefined;
+}
+
+async function defaultHandlePostSignUp(user: User, context: TypeContextEmailPassowrd | TypeContextThirdParty) {}
+
+async function defaultHandlePostSignIn(user: User, context: TypeContextEmailPassowrd | TypeContextThirdParty) {}
+
+async function defaultSetSessionDataForSession(
+    user: User,
+    context: TypeContextEmailPassowrd | TypeContextThirdParty,
+    action: "signin" | "signup"
+) {
+    return {};
+}
+
+async function defaultSetJwtPayloadForSession(
+    user: User,
+    context: TypeContextEmailPassowrd | TypeContextThirdParty,
+    action: "signin" | "signup"
+) {
+    return {};
+}
+
+function validateAndNormaliseSignUpConfig(
+    recipeInstance: Recipe,
+    appInfo: NormalisedAppinfo,
+    config?: TypeInputSignUp
+): TypeNormalisedInputSignUp {
+    let disableDefaultImplementation =
+        config === undefined || config.disableDefaultImplementation === undefined
+            ? false
+            : config.disableDefaultImplementation;
+
+    let formFields: NormalisedFormField[] = [];
+    if (config !== undefined && config.formFields !== undefined) {
```

Comment:
Why is this not using the normalising function from emailpassword itself? Why duplicate code..

Reply:
done